src/PartitionLoops.cpp

```diff
@@ -723,21 +723,21 @@ class RenormalizeGPULoops : public IRMutator {
             return;
         }
 
-        if (ends_with(op->name, "__thread_id_x")) {
-            in_thread_loop = true;
-            IRMutator::visit(op);
-            in_thread_loop = false;
-            return;
-        }
-
         bool old_in_gpu_loop = in_gpu_loop;
 
         if (in_gpu_loop || CodeGen_GPU_Dev::is_gpu_var(op->name)) {
             gpu_vars.push(op->name, 0);
             in_gpu_loop = true;
         }
 
-        IRMutator::visit(op);
+
+        if (ends_with(op->name, "__thread_id_x")) {
+            in_thread_loop = true;
+            IRMutator::visit(op);
+            in_thread_loop = false;
+        } else {
+            IRMutator::visit(op);
```

Comment:
I think this pattern reads better if there is one visit call and the state setting happens around it. Ideally we'd have an RAII class for this pattern so it would be:
    SaveRestoreFlag scope(in_thread_loop, ends_with(op->name, "__thread_id_x"));